ROX-21571,ROX-21605: Fix Splunk integration tests

Maddosaurus opened the pull request:
### Description

Most of the flakes in the two tickets were caused by image pull errors.
With the addition of the image prefetcher, these errors should not happen anymore. I've verified that all images used in the test are covered by the prefetcher.

In the investigation I noticed that we never tested the Splunk integration with modern Splunk, as this was hardcoded in the test:  `SplunkUtil.createSplunk(orchestrator, ORCHESTRATOR_NAMESPACE, true` instead of using the variable `legacy`.
This led to additional changes, as Splunk 9 requires the creation of a token instead of shipping a default one.
I also improved the general robustness of queries to find data.

### Review Pointers
Following the logical changes of the PR might make for a more pleasant review.
- Discover the hardwired var in [IntegrationsTest.groovy](https://github.com/stackrox/stackrox/pull/12090/files#diff-76981cd6624827e4694cb4978c4de7270547bf5926d0206a17e90ec13c9764a5)
- See the added logic for HEC tokens as well as the improvements of search queries in [Notifiers.groovy](https://github.com/stackrox/stackrox/pull/12090/files#diff-866113a9cf6a12315fc4aa9c0e5b54fae2f1b557b758a9949b8ecbe0333461dd)
- Discover cascading changes and improvements in the remaining files, which hold the utility methods for aforementioned changes

### User-facing documentation

- [x] CHANGELOG update is not needed
- [x] Documentation is not needed

### Testing

- [ ] inspected CI results

#### Automated testing
- [x] modified existing tests


#### How I validated my change
- Manual testing against an OpenShift infra Cluster with `./gradlew test --tests="IntegrationsTest.*Splunk*"`
- CI

Maddosaurus commented on qa-tests-backend/src/main/groovy/objects/Notifiers.groovy:
```diff
@@ -136,8 +142,9 @@ class SplunkNotifier extends Notifier {
 
     void validateViolationNotification(Policy policy, Deployment deployment, boolean strictIntegrationTesting) {
-        def response = SplunkUtil.waitForSplunkAlerts(splunkPort, 30)
+        def response = SplunkUtil.waitForSplunkAlerts(splunkPort, "search sourcetype=stackrox-alert")
```
Search for specifically the deployment violation for given deploymentUid

Maddosaurus commented:
@rhacs-bot run gke-qa-e2e-tests

Maddosaurus commented:
/test gke-qa-e2e-tests

vikin91 commented on qa-tests-backend/src/main/groovy/objects/Notifiers.groovy:
```diff
@@ -124,7 +124,13 @@ class SplunkNotifier extends Notifier {
+        String hecToken  = "00000000-0000-0000-0000-000000000000" // default for Splunk 6
+        if (!legacy) {
+            // For current Splunk images an ingest token needs to be created
+            log.info("Creating HEC ingest token")
+            hecToken = SplunkUtil.createHECToken(splunkPort)
+        }
```
```suggestion
            hecToken = SplunkUtil.createHECToken(splunkPort)
        }
        log.info("Using HEC ingest token: ${hecToken}")
```

vikin91 replied:
Do we want to assert that `SplunkUtil.createHECToken(splunkPort)` returned something reasonable?

Maddosaurus replied:
Yes, I've added an assert to `createHECToken` to ensure the token has the expected length

vikin91 commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -141,6 +133,26 @@ class SplunkUtil {
         }
     }
 
+    static String createHECToken(int port, String tokenName = "stackrox") {
+        Response response = null
+        withRetry(20, 15) {
+            response =  given().auth()
+                .basic("admin", SPLUNK_ADMIN_PASSWORD)
+                .formParam("name", tokenName)
+                .formParam("output_mode", "json")
+                .post("https://127.0.0.1:${port}/servicesNS/nobody/search/data/inputs/http")
+        }
+        return unmarshalHEC(response?.asString())
+    }
```
```suggestion
    static String createHECToken(int port, String tokenName = "stackrox") {
        Response response = null
        withRetry(20, 15) {
            log.info("Attempting to create new HEC ingest token")
            response =  given().auth()
                .basic("admin", SPLUNK_ADMIN_PASSWORD)
                .formParam("name", tokenName)
                .formParam("output_mode", "json")
                .post("https://127.0.0.1:${port}/servicesNS/nobody/search/data/inputs/http")
        }
        return unmarshalHEC(response?.asString())
    }
```


This would make the logs a bit easier to understand. Currently they print:

```
    15:02:54 | INFO  | IntegrationsTest          | SplunkNotifier            | Creating HEC ingest token
    15:02:54 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 0 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:03:09 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 1 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:03:24 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 2 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:03:39 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 3 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:03:54 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 4 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:04:10 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 5 of 20): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:04:27 | DEBUG | IntegrationsTest          | SplunkNotifier            | validating splunk deployment is ready to accept events before creating notifier...
```

I think, it would be cool to know that the handshake errors are coming from the  `createHECToken` function.

vikin91 replied:
Note that after the `validating splunk deployment is ready to accept events before creating notifier...` there will be another group of similar errors repeated only 6 times - it would be cool to add more logging about where those errors are originating from.

For `legacy=true`, we see:
```
    15:07:47 | DEBUG | IntegrationsTest          | SplunkNotifier            | validating splunk deployment is ready to accept events before creating notifier...
    15:07:48 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 0 of 6): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:08:03 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 1 of 6): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:08:18 | DEBUG | IntegrationsTest          | Helpers                   | Caught exception. Retrying in 15s (attempt 2 of 6): javax.net.ssl.SSLHandshakeException: Remote host terminated the handshake
    15:08:34 | DEBUG | IntegrationsTest          | SplunkUtil                | {"sid":"1721747314.1"}
    15:08:34 | DEBUG | IntegrationsTest          | SplunkUtil                | New Search created: 1721747314.1
```

My questions here are:
- Is 6 repetitions enough for legacy=true assuming that legacy=false sometimes needs 11 out of planned 30?
- Can we maybe wait separately for Splunk to be ready and accept the handshake and then do the bussinnes logic (i.e, create token or creating notifier)?

Maddosaurus replied:
Reading this test again, I think it makes more sense to have a `waitForSplunkBoot` (or something similar) that is called after the container creation. I've already identified a [good API candidate for that](https://docs.splunk.com/Documentation/Splunk/7.3.1/RESTREF/RESTintrospect#server.2Fhealth.2Fsplunkd).
That way, we could have clean logs and do away with this workaround of running a `createSearch` just to ensure that Splunk is ready.
That should also result in these API calls not erroring out with an exception multiple times.

vikin91 replied:
Do you think, you could add the `waitForSplunkBoot` to this PR? Or is it too much for now?

Maddosaurus replied:
Yes, I added this as utility function in [02e341c](https://github.com/stackrox/stackrox/pull/12090/commits/02e341c86271641fed5871bb4009c8926a1f452d).
I will follow up with a separate PR to update the the other Splunk groovy test (SplunkIntegrationTest) with this new function.

vikin91 replied:
Okay, but let's please not postpone this follow-up too much if possible. I wish, we do our best on the Splunk test and hope to not see them flake for a long time.

vikin91 reviewed (commented):
The changes to the test look good! 
I think, we could add few improvements - see comments.

lvalerom commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -141,6 +133,26 @@ class SplunkUtil {
+    static String createHECToken(int port, String tokenName = "stackrox") {
+        Response response = null
+        withRetry(20, 15) {
+            response =  given().auth()
+                .basic("admin", SPLUNK_ADMIN_PASSWORD)
+                .formParam("name", tokenName)
+                .formParam("output_mode", "json")
+                .post("https://127.0.0.1:${port}/servicesNS/nobody/search/data/inputs/http")
```
I think we should assert that the response is valid here.

Maddosaurus replied:
Good idea. Added a check for token length

lvalerom commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -141,6 +133,26 @@ class SplunkUtil {
+                .formParam("output_mode", "json")
+                .post("https://127.0.0.1:${port}/servicesNS/nobody/search/data/inputs/http")
+        }
+        return unmarshalHEC(response?.asString())
```
It would be good to assert here if `unmarshalHEC` returned a valid token. Maybe this becomes redundant if we already asserted that the response is valid earlier.

Maddosaurus replied:
I assert on the return of `unmarshalHEC`, as this is the function that retrieves the raw data

lvalerom commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -141,6 +133,26 @@ class SplunkUtil {
         }
     }
 
+    static String createHECToken(int port, String tokenName = "stackrox") {
```
would it make sense to pass the `collectorServiceName` here and not hardcode `127.0.0.1`?

Maddosaurus replied:
Given that we do a port forward to localhost, I think using a hardcoded value is fine

Maddosaurus commented:
@vikin91 Follow-up PR for the other Splunk integration test: https://github.com/stackrox/stackrox/pull/12138

vikin91 commented:
> @vikin91 Follow-up PR for the other Splunk integration test: #12138

Thanks! I am still wondering whether this PR would not be considered flaky without #12138...

I'll wait for the CI to run this test at least once before giving the ✅ if that is okay

Maddosaurus commented:
@vikin91 
> Thanks! I am still wondering whether this PR would not be considered flaky without https://github.com/stackrox/stackrox/pull/12138...

Can you elaborate on that? This PR (12090) tests the push integration from StackRox to Splunk.
12138 tests the pull integration for Splunk.
While both benefit from the improvements, they are testing two distinct integrations, so I'd argue that they don't have anything in common except the logic for handling a Splunk deployment.

> I'll wait for the CI to run this test at least once before giving the ✅ if that is okay

Yes, absolutely. I'll also give it at least 5 re-runs.
Do you happen to remember how to get rhacs-bot to do that?

vikin91 commented:
> Can you elaborate on that? This PR (12090) tests the push integration from StackRox to Splunk. 12138 tests the pull integration for Splunk. While both benefit from the improvements, they are testing two distinct integrations, so I'd argue that they don't have anything in common except the logic for handling a Splunk deployment.


Oh, now I see that those are two different tests. This makes sense ;) My comment was written as if the #12138 would be the improvement to the same test as the one already being improved in this PR.

> Yes, absolutely. I'll also give it at least 5 re-runs. Do you happen to remember how to get rhacs-bot to do that?

You need the label `auto-retest` and some `/test` command, I think it was `/retest-times 10 gke-qa-e2e-tests` but please double check that as I can't remember the proper syntax.

Maddosaurus commented:
The new `waitForSplunkBoot` failed ([link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/stackrox_stackrox/12090/pull-ci-stackrox-stackrox-master-gke-qa-e2e-tests/1816396816418082816)) because of the SPlunk overall status not being green but red instead because of IOWait.
IOWait warnings happen quite quickly, because we are not considerate with wait times (e.g. re-trigger search queries or setting low scrape intervals). I will rework the check to only focus on the Index Processor and the Search Scheduler, as only these are of interest to us.
Here is the relevant part of the status json - I intend to check the category-wide "health" entry:
```json
{

    "entry": [
        {
            "name": "splunkd",
            "content": {
                "eai:acl": null,
                "features": {
                    "File Monitor Input": {
                        "features": {
                            "Forwarder Ingestion Latency": {
                                "health": "green"
                            },
                            "Ingestion Latency": {
                                "health": "green"
                            },
                            "Large and Archive File Reader-0": {
                                "health": "green"
                            },
                            "Real-time Reader-0": {
                                "health": "green"
                            }
                        },
                        "health": "green"
                    },
                    "Index Processor": {
                        "features": {
                            "Buckets": {
                                "health": "green"
                            },
                            "Disk Space": {
                                "health": "green"
                            },
                            "Index Optimization": {
                                "health": "green"
                            }
                        },
                        "health": "green"
                    },
                    "Search Scheduler": {
                        "features": {
                            "Scheduler Suppression": {
                                "health": "green"
                            },
                            "Search Lag": {
                                "health": "green"
                            },
                            "Searches Delayed": {
                                "health": "green"
                            },
                            "Searches Skipped in the last 24 hours": {
                                "health": "green"
                            }
                        },
                        "health": "green"
                    },
                    "Workload Management": {
                        "disabled": true,
                        "health": "green"
                    }
                },
                "health": "green"
            }
        }
    ],
    "paging": {
        "total": 1,
        "perPage": 30,
        "offset": 0
    },
    "messages": []
}
```

Maddosaurus commented:
I've decided to remove Splunk 6 capability, as it has been EOL / out of support for more than 4 years now.
This simplifies setup and config of the Splunk deployments and cuts down on test times for this integration test.
(https://www.splunk.com/en_us/legal/splunk-software-support-policy.html)

Maddosaurus commented:
/retest-times

Maddosaurus commented:
/retest-times 2 gke-qa-e2e-tests

Maddosaurus commented:
/test gke-qa-e2e-tests

Maddosaurus commented:
/test gke-qa-e2e-tests

Maddosaurus commented:
Note: The test that has been failing in the e2e tests is **not** this IntegrationTest I have updated in this PR, but instead the **other** SplunkIntegrationTest - that will be fixed by https://github.com/stackrox/stackrox/pull/12138. 
The other PR is based on this branch and will be merged once this PR here is merged.

vikin91 commented:
I see there are some groovy errors in the log..

```
groovy.lang.MissingMethodException: No signature of method: static util.SplunkUtil.createSplunk() is applicable for argument types: (orchestratormanager.Kubernetes, String, Boolean) values: [orchestratormanager.Kubernetes@4bcc57de, qa-splunk, false]
Possible solutions: createSplunk(orchestratormanager.OrchestratorMain, java.lang.String)
	at IntegrationsSplunkViolationsTest.setup(IntegrationsSplunkViolationsTest.groovy:59) [3 skipped]
	at util.OnFailureInterceptor.intercept(OnFailure.groovy:72) [9 skipped]
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:148) [13 skipped]
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at util.OnFailureInterceptor.intercept(OnFailure.groovy:72) [6 skipped]
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61) [4 skipped]
 [4 skipped]
```

I think, this could be fixed, right?

Maddosaurus commented:
Yes, I've changed the signature of that function. That is already [included in the other PR](https://github.com/stackrox/stackrox/pull/12138/files#diff-b5d8e8c78f856dbec06efbf9fdf975540853d6691648c5071a853559e9e47a5bL59) but for passing these tests correctly, I've also added it here in [6379669](https://github.com/stackrox/stackrox/pull/12090/commits/6379669628ccf02fe2b6609b4c5bb269b34548ca)

Maddosaurus commented:
/retest gke-sensor-integration-tests

Maddosaurus commented:
/test gke-sensor-integration-tests

vikin91 commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -62,26 +59,16 @@ class SplunkUtil {
-            }
+        withRetry(40, 15) {
+            def searchId
+            searchId = createSearch(port, search)
```
Question: do we need to create a new search on every retry?
Would it be possible to create the search once and then run `getSplunkAlerts` in the retry-loop?

Maddosaurus replied:
From what I can find in the API docs, a new search needs to be created for every try

vikin91 commented on qa-tests-backend/src/main/groovy/util/SplunkUtil.groovy:
```diff
@@ -193,6 +201,30 @@ class SplunkUtil {
+                // as overall health can be degraded because of IOWait or disk storage.
+                assert entry.content.features.searchScheduler.health == "green"
+                // The IndexProcessor can be yellow because of free disk space. That's okay for the short-lived test.
+                assert (entry.content.features.indexProcessor.health == "green" ||
+                        entry.content.features.indexProcessor.health == "yellow")
```
Fully optional:
```suggestion
                assert (entry.content.features.indexProcessor.health != "red")
```

vikin91 reviewed (approved):
I gave it another look and it looks good! 
I am leaving two optional comments/questions - feel free to decide whether you want to address them and resolve the conversations.

Maddosaurus commented:
/test gke-qa-e2e-tests
